service/worker/scanner/executions/mutable_state_validator.go

```diff
@@ -207,9 +233,42 @@ func (v *mutableStateIDValidator) validateSignal(
 	return results
 }
 
-func (v *mutableStateIDValidator) validateID(
+func (v *mutableStateValidator) validateID(
 	eventID int64,
 	lastEventID int64,
 ) bool {
 	return common.FirstEventID <= eventID && eventID <= lastEventID
 }
+
+func (v *mutableStateValidator) validateRetention(
+	executionInfo *persistencespb.WorkflowExecutionInfo,
+	executionState enums.WorkflowExecutionState,
+) (*MutableStateValidationResult, error) {
+	if executionState != enums.WORKFLOW_EXECUTION_STATE_COMPLETED {
+		return nil, nil
+	}
+	// We don't use the close time here because some old workflows do not have the close time.
+	// It makes sense the workflow is finished and use the last update time.
+	finalUpdateTime := executionInfo.GetLastUpdateTime()
+	if finalUpdateTime == nil {
+		return nil, nil
```

Comment:
I think this is not expected? Maybe add a warning log here?

Reply:
Return a internal error instead.